File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/FeedHTTPTransport.java
package org.nypl.simplified.books.core;

import com.io7m.jfunctional.OptionType;
import com.io7m.jnull.NullCheck;
import org.nypl.simplified.http.core.HTTPAuthType;
import org.nypl.simplified.http.core.HTTPRedirectFollower;
import org.nypl.simplified.http.core.HTTPResultError;
import org.nypl.simplified.http.core.HTTPResultException;
import org.nypl.simplified.http.core.HTTPResultMatcherType;
import org.nypl.simplified.http.core.HTTPResultOKType;
import org.nypl.simplified.http.core.HTTPResultType;
import org.nypl.simplified.http.core.HTTPType;
import org.nypl.simplified.opds.core.OPDSFeedTransportException;
import org.nypl.simplified.opds.core.OPDSFeedTransportIOException;
import org.nypl.simplified.opds.core.OPDSFeedTransportType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;

/**
 * An implementation of the {@link OPDSFeedTransportType} interface that uses an
 * {@link HTTPType} instance for communication, supporting optional
 * authentication.
 */

public final class FeedHTTPTransport
  implements OPDSFeedTransportType<OptionType<HTTPAuthType>>
{
  private static final Logger LOG;

  static {
    LOG = NullCheck.notNull(LoggerFactory.getLogger(FeedHTTPTransport.class));
  }

  private final HTTPType http;

  private FeedHTTPTransport(final HTTPType in_http)
  {
    this.http = NullCheck.notNull(in_http);
  }

  /**
   * @param http An HTTP interface
   *
   * @return A new transport
   */

  public static OPDSFeedTransportType<OptionType<HTTPAuthType>> newTransport(
    final HTTPType http)
  {
    return new FeedHTTPTransport(http);
  }

  @Override public InputStream getStream(
    final OptionType<HTTPAuthType> auth,
    final URI uri)
    throws OPDSFeedTransportException
  {
    final HTTPRedirectFollower rf = new HTTPRedirectFollower(
      FeedHTTPTransport.LOG, this.http, auth, 5, uri, 0L);

    final HTTPResultType<InputStream> r = rf.run();
    return r.matchResult(
      new HTTPResultMatcherType<InputStream, InputStream,
        OPDSFeedTransportException>()
      {
        @Override
        public InputStream onHTTPError(final HTTPResultError<InputStream> e)
          throws OPDSFeedTransportException
        {
          throw new FeedTransportHTTPException(
            e.getMessage(), e.getStatus(), e.getProblemReport());
        }

        @Override public InputStream onHTTPException(
          final HTTPResultException<InputStream> e)
          throws OPDSFeedTransportException
        {
          final Exception er = e.getError();
          final IOException ex = new IOException(er);
          throw new OPDSFeedTransportIOException(er.getMessage(), ex);
        }

        @Override
        public InputStream onHTTPOK(final HTTPResultOKType<InputStream> e)
          throws OPDSFeedTransportException
        {
          return e.getValue();
        }
      });
  }
}

File: simplified-feeds-api/src/main/java/org/nypl/simplified/feeds/api/FeedHTTPTransport.kt
package org.nypl.simplified.feeds.api

import one.irradia.mime.api.MIMECompatibility
import org.librarysimplified.http.api.LSHTTPClientType
import org.librarysimplified.http.api.LSHTTPRequestBuilderType
import org.librarysimplified.http.api.LSHTTPResponseStatus
import org.nypl.simplified.accounts.api.AccountReadableType
import org.nypl.simplified.accounts.api.setAuthentication
import org.nypl.simplified.opds.core.OPDSFeedTransportException
import org.nypl.simplified.opds.core.OPDSFeedTransportIOException
import org.nypl.simplified.opds.core.OPDSFeedTransportType
import org.slf4j.LoggerFactory
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.net.URI
import java.util.Locale

/**
 * An implementation of the [OPDSFeedTransportType] interface that uses an
 * [HTTPType] instance for communication, supporting optional
 * authentication.
 */

class FeedHTTPTransport(
  private val http: LSHTTPClientType
) : OPDSFeedTransportType<AccountReadableType> {

  private val logger =
    LoggerFactory.getLogger(FeedHTTPTransport::class.java)

  @Throws(OPDSFeedTransportException::class)
  override fun getStream(
    account: AccountReadableType,
    uri: URI,
    method: String,
    authenticate: Boolean
  ): InputStream {
    this.logger.debug("get stream: {} {}", uri, account)

    val request =
      this.http.newRequest(uri)
        .setMethod(this.methodOfName(method))
        .apply { if (authenticate) { setAuthentication(account) } }
        .build()

    val response = request.execute()
    return when (val status = response.status) {
      is LSHTTPResponseStatus.Responded.OK ->
        status.bodyStream ?: ByteArrayInputStream(ByteArray(0))

      is LSHTTPResponseStatus.Responded.Error ->
        throw FeedHTTPTransportException(
          message = status.properties.message,
          code = status.properties.status,
          report = status.properties.problemReport
        )

      is LSHTTPResponseStatus.Failed ->
        throw OPDSFeedTransportIOException(
          message = "Connection failed",
          cause = IOException(status.exception)
        )
    }
  }

  private fun methodOfName(method: String): LSHTTPRequestBuilderType.Method {
    return when (method.toUpperCase(Locale.ROOT)) {
      "GET" -> LSHTTPRequestBuilderType.Method.Get
      "HEAD" -> LSHTTPRequestBuilderType.Method.Head
      "POST" -> LSHTTPRequestBuilderType.Method.Post(ByteArray(0), MIMECompatibility.applicationOctetStream)
      "PUT" -> LSHTTPRequestBuilderType.Method.Put(ByteArray(0), MIMECompatibility.applicationOctetStream)
      "DELETE" -> LSHTTPRequestBuilderType.Method.Delete
      else -> throw IllegalArgumentException("Unsupported request method: $method")
    }
  }
}
